### ISSUE TITLE
Crash when yielding to a done task

### ISSUE THREAD

**Issue by lbilli** [JuliaLang/julia#36866]
Julia 1.5.0-rc2 on Linux:
```julia
julia> t = @task begin end

julia> yield(t)

julia> yield(t)
ERROR: fatal: error thrown and no exception handler available.
TypeError(func=:typeassert, context="", expected=Int32, got=nothing)
jl_type_error_rt at /buildworker/worker/package_linux64/build/src/rtutils.c:118
jl_type_error at /buildworker/worker/package_linux64/build/src/rtutils.c:125
uv_write at ./stream.jl:933
unsafe_write at ./stream.jl:1005
write at ./strings/io.jl:183 [inlined]
print at ./strings/io.jl:185 [inlined]
#with_output_color#745 at ./util.jl:90
with_output_color##kw at ./util.jl:70 [inlined]
#printstyled#746 at ./util.jl:104 [inlined]
printstyled##kw at ./util.jl:104
unknown function (ip: 0x7f2f87037ffc)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2231 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
display_error at ./client.jl:102
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2231 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
display_error at ./client.jl:106
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2231 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1690 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:655
jl_f__apply_latest at /buildworker/worker/package_linux64/build/src/builtins.c:705
#invokelatest#1 at ./essentials.jl:710 [inlined]
invokelatest at ./essentials.jl:709 [inlined]
_start at ./client.jl:508
```

**Comment by vtjnash** [comment]
Ah, I've always suspected this code path was bad, but didn't know how to trigger it.
```diff
diff --git a/src/task.c b/src/task.c
index fd93695b97..9ffd06cf19 100644
--- a/src/task.c
+++ b/src/task.c
@@ -390,9 +390,7 @@ JL_DLLEXPORT void jl_switch(void)
     }
     if (t->state == done_sym || t->state == failed_sym ||
             (t->started && t->stkbuf == NULL)) {
-        ct->exception = t->exception;
-        ct->result = t->result;
-        return;
+        jl_error("cannot switch to dead task");
     }
     if (ptls->in_finalizer)
         jl_error("task switch not allowed from inside gc finalizer");
```

**Comment by JeffBezanson** [comment]
Arguably there is also a problem with `yield(t)` (to the extent that should exist at all), where it enqueues the current task before checking anything about the target task.

**Comment by lbilli** [comment]
Fixed in Julia 1.7.0 by #41410
